docs/protocol-initialization-facet.md

```diff
@@ -0,0 +1,57 @@
+
+Facets allows to query the current version by calling `getVersion`.
+
+### Facet initialization
+
+Each Protocol facet should have initialization function (even if no-op) to ensure consistency across codebase and different version. Initialization function should be written as if protocol is deployed for the first time (i.e. not taking into account potential effects on the upgrade). Although initialization functions must be external, they should not be part of facet interface and therefore should never be added to diamond function list. This is important to prevent facet reinitialization and clashes between initializers with same argument types.
+
+When protocol is deployed for the first time, `initialize` on Protocol initialization handler should get list of all facet implementation addresses together with corresponding initialization data. This data is passed directly on individual facets where whole initialization takes place.  
+
+However, before upgrade is done, effects of facet initializers should be carefully considered before passing data to initialization handler. If facet's initialization function does not affect the protocol state in a harmful way, it should be passed to initialization handler in the same way as for the initial deployment. However, when initialize affects the storage (for example Config handler sets all counters to 1), facet should be omitted from initialization call and approach from next section should be followed.
+
+### Data initialization
+
+When facet initializers clash with existing data or simply a specific storage must be populated, initializer facet must be updated to cater for this kind of change. Suppose new version is X.Y.Z and there is a need to modify the storage during the update, a new internal function should be prepared, called `initVX_Y_Z`. This function can make additional version checks and can accept arbitrary data which can be handled in any desired way. This allows populating custom storage slots during the upgrade or mimicking actions of the facet initializers that would otherwise be harmful. For example if another limit is added to the Config handler, `initVX_Y_Z` could simply store new value to desired location, without the need to overwrite other config values and at the same time leaves counters intact.  
+Initialization data is passed in as bytes, so `initVX_Y_Z` must be decoded into correct types if needed.
+
+### Managing supported interfaces
+
+Although most facets initializers automatically register their EIP165 interfaces, old interfaces are not automatically removed during upgrade/removal. Protocol initialization handler therefore allows upgrader to supply list of interfaces to remove or add, which happens atomically during the upgrade.
+
+
+### Initialization diagram
+Diagram below represents a simple upgrade with two upgraded facets:
```

Comment:
The diagram represents a simple protocol upgrade whereby two facets are upgraded:

Reply:
Done.